### PR TITLE
Using MapData for Tasks

### LINKED ISSUE #202: TaskType currently not known for tasks
Since https://github.com/Impostor/Impostor/pull/164 fixed the index being used as TaskTypes, we don't have any information about the TaskTypes in memory now.

The `TaskInfo.Type` property should be properly updated to include the proper `TaskTypes`.

### LINKED ISSUE #261: Complete Task Suggestion
# Api Suggestion
## Suggestion
Allow completing tasks from specific players. (currently it's `get;` only)  
eg.: `IInnerPlayerInfo#Tasks.ToList()[0].SetCompleteAsync(bool);`

## Use cases
• Programmatically complete tasks for a single player;  
• Simulate a timer in the taskbar.

### PR DESCRIPTION
**Pull request by Minorusama**
### Description
Rework of #275.

- Adds a method to complete a task in TaskInfo
- Adds a method to complete all tasks in InnerPlayerControl
- Adds a refence to ITask in TaskInfo to identify the task

### Closes issues

- closes #261
- closes #202

### CODE REVIEW

**Comment by JohnTranQUT** [comment]
@Minorusama I've tested and found that the tasks for AirShip was off.
I've made the fixes and left in https://github.com/Minorusama/Impostor/pull/4 pull for you to update.

- [x] Also gotten rid of a warning
- [x] Add in an example event-listener.
- [ ] Other maps tasks tested thoroughly (if innersloth changed anything).

**Review by js6pak — CHANGES_REQUESTED** [review]

File: src/Impostor.Api/Net/Inner/Objects/IInnerPlayerControl.cs [hunk 1]
```diff
@@ -105,5 +105,12 @@ public interface IInnerPlayerControl : IInnerNetObject
         /// </summary>
         /// <returns>Task that must be awaited.</returns>
         ValueTask ExileAsync();
+
+        /// <summary>
+        ///     Sets all the player's tasks as complete, if the player is a crewmate.
+        ///     Visible to all players.
+        /// </summary>
+        /// <returns>Task that must be awaited.</returns>
+        ValueTask CompleteAllTasksAsync();
```
[the remark this answers is not in the document]
Comment (miniduikboot) [reply src/Impostor.Api/Net/Inner/Objects/IInnerPlayerControl.cs:114]: Agree, I don't think this method belongs in the public API

File: src/Impostor.Plugins.Example/Handlers/PlayerEventListener.cs [hunk 2]
```diff
@@ -84,6 +84,11 @@ public async ValueTask OnPlayerChat(IPlayerChatEvent e)
                 await e.PlayerControl.NetworkTransform.SnapToAsync(new Vector2(1, 1));
             }
 
+            if (e.Message == "completetasks")
+            {
+                await e.PlayerControl.CompleteAllTasksAsync();
```
Comment (miniduikboot) [inline src/Impostor.Plugins.Example/Handlers/PlayerEventListener.cs:89]: just inline CompleteAllTasksAsync here and remove it from the public api

File: src/Impostor.Server/Net/Inner/Objects/InnerGameData.TaskInfo.Api.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+                if (_playerInfo.Controller == null)
+                {
+                    throw new ImpostorException("Can't complete a task that doesn't have a player assigned");
+                }
+
+                var player = _playerInfo.Controller;
+
+                if (Complete)
+                {
+                    return;
```
Comment (miniduikboot) [inline src/Impostor.Server/Net/Inner/Objects/InnerGameData.TaskInfo.Api.cs:24]: do we want to throw an exception when api users try to complete an already completed tasks? The method not doing anything when it is already completed would surprise me

File: src/Impostor.Server/Net/Inner/Objects/InnerPlayerInfo.cs [hunk 4]
```diff
@@ -56,7 +58,7 @@ public void Serialize(IMessageWriter writer)
             throw new NotImplementedException();
         }
 
-        public void Deserialize(IMessageReader reader)
+        public void Deserialize(IMessageReader reader, IEventManager eventManager, Game game)
```
Comment (miniduikboot) [inline src/Impostor.Server/Net/Inner/Objects/InnerPlayerInfo.cs:61]: what are these new arguments used for?

**Review by js6pak — APPROVED** [review]

**Review by miniduikboot — APPROVED** [review]

**Pull request merged**